relayer/relayer/src/trader/mod.rs

```diff
@@ -0,0 +1,303 @@
+use std::{
+    fmt::Debug,
+    str::FromStr,
+    sync::Arc,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use aleph_client::{pallets::system::SystemApi, AccountId, AsConnection, SignedConnectionApi};
+use ethers::providers::Middleware;
+use log::{debug, error, info, warn};
+use thiserror::Error;
+use tokio::{select, sync::broadcast, time::sleep};
+
+use crate::{
+    config::Config,
+    connections::{azero::AzeroConnectionWithSigner, eth::SignedEthConnection},
+    contracts::{AzeroContractError, AzeroEtherInstance, MostInstance, RouterInstance},
+    helpers::left_pad,
+    CircuitBreakerEvent,
+};
+
+// trader component will sell the surplus
+pub const ONE_AZERO: u128 = 1_000_000_000_000;
+pub const ONE_ETHER: u128 = 1_000_000_000_000_000_000;
+
+pub const ETH_TO_AZERO_RELAYING_BUFFER: u128 = 100 * ONE_AZERO;
+pub const TRADED_AZERO_FEE_MULTIPLIER: u128 = 20;
+pub const SLIPPAGE_PERCENT: u128 = 1;
+
+pub const HOUR_IN_MILLIS: u64 = 60 * 60 * 1000;
+pub const TRADER_QUERY_INTERVAL_MILLIS: u64 = 30 * 1000;
+
+#[derive(Debug, Error)]
+#[error(transparent)]
+#[non_exhaustive]
+pub enum TraderError {
+    #[error("Error when parsing ethereum address {0}")]
+    FromHex(#[from] rustc_hex::FromHexError),
+
+    #[error("AlephClient error {0}")]
+    AlephClient(#[from] anyhow::Error),
+
+    #[error("Azero contract error {0}")]
+    AzeroContract(#[from] AzeroContractError),
+
+    #[error("Broadcast receive error {0}")]
+    BroadcastReceive(#[from] broadcast::error::RecvError),
+
+    #[error("Missing required arg {0}")]
+    MissingRequired(String),
+
+    #[error("Address is not an AccountId {0}")]
+    NotAccountId(String),
+
+    #[error("Trader has exited unexpectedly - this should never happen")]
+    TraderExited,
+}
+
+#[derive(Copy, Clone)]
+pub struct Trader;
```

Comment:
Question: why is this a struct if the only big method does not take `&self` and it contains no data? You could just have a function instead. But see below.